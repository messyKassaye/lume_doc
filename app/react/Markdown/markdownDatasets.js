import HtmlParser from 'htmlparser2/lib/Parser';
import queryString from 'query-string';
import rison from 'rison-node';
import Big from 'big.js';

import searchApi from 'app/Search/SearchAPI';
import api from 'app/utils/api';
import entitiesApi from 'app/Entities/EntitiesAPI';

const conformUrl = ({ url = '', geolocation = false }) => {
  const { q } = queryString.parse(url.substring(url.indexOf('?')));

  if (!q) {
    const defaultValue = geolocation
      ? { allAggregations: true, limit: 0, geolocation: true }
      : { allAggregations: true, limit: 0 };

    return defaultValue;
  }

  const params = rison.decode(q);
  params.limit = 0;

  if (geolocation) {
    params.geolocation = true;
  }

  return params;
};

const conformValues = attribs => (attribs.entity ? attribs : conformUrl(attribs));

const parseDatasets = markdown => {
  const result = {};
  const parser = new HtmlParser(
    {
      onopentag(name, attribs) {
        if (name === 'dataset') {
          result[attribs.name || 'default'] = conformValues(attribs);
        }
        if (name === 'query') {
          result[attribs.name || 'default'] = { url: attribs.url, query: true };
        }
      },
    },
    { decodeEntities: true }
  );

  parser.parseComplete(markdown);
  return result;
};

const requestDatasets = (datasets, requestParams) =>
  Promise.all(
    Object.keys(datasets).map(name => {
      if (datasets[name].query) {
        return api.get(datasets[name].url, requestParams).then(data => ({ data: data.json, name }));
      }
      const apiAction = datasets[name].entity ? entitiesApi.get : searchApi.search;
      const params = datasets[name].entity ? { sharedId: datasets[name].entity } : datasets[name];
      const postAction = datasets[name].entity ? d => d[0] : d => d;
      return apiAction(requestParams.set(params))
        .then(postAction)
        .then(data => ({ data, name }));
    })
  );

const conformDatasets = sets => sets.reduce((memo, set) => ({ ...memo, [set.name]: set.data }), {});

const getAggregations = (state, { property, dataset = 'default' }) => {
  const data = state.page.datasets.get(dataset);
  return !data ? undefined : data.getIn(['aggregations', 'all', property, 'buckets']);
};

const addValues = (aggregations, values) => {
  let result = new Big(0);
  values.forEach(key => {
    const value = aggregations.find(bucket => bucket.get('key') === key);
    const filteredValue = value ? value.getIn(['filtered', 'doc_count']) : 0;
    result = result.plus(filteredValue || 0);
  });
  return Number(result);
};

export default {
  async fetch(markdown, requestParams) {
    const datasets = parseDatasets(markdown);
    return requestDatasets(datasets, requestParams).then(conformDatasets);
  },

  getRows(state, { dataset = 'default' }) {
    const data = state.page.datasets.get(dataset);
    if (!data) {
      return undefined;
    }
    return data.get('rows');
  },

  getAggregations,

  getAggregation(state, { uniqueValues, property, value, dataset = 'default' }) {
    const aggregations = getAggregations(state, { property, dataset });
    if (!aggregations) {
      return undefined;
    }

    if (uniqueValues) {
      return aggregations.filter(a => a.getIn(['filtered', 'doc_count']) !== 0).size;
    }

    const values = value ? value.split(',') : [''];
    return addValues(aggregations, values);
  },

  getMetadataValue(state, { property, dataset = 'default' }) {
    const data = state.page.datasets.get(dataset);
    const propertyExists = data && data.hasIn(['metadata', property]);
    const mos = propertyExists ? data.getIn(['metadata', property]).toJS() : [];
    return mos && mos.length && mos[0].value ? Number(mos[0].value) : undefined;
  },
};
