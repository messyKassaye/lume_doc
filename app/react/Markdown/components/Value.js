import React, { Component } from 'react';
import { connect } from 'react-redux';
import PropTypes from 'prop-types';

import markdownDatasets from '../markdownDatasets';
import PagesContext from './Context';
import { objectPath } from '../utils';

export class ValueComponent extends Component {
  renderChildren(_value) {
    const { path } = this.props;
    return path ? objectPath(path, _value) : _value;
  }

  render() {
    const { property, value } = this.props;
    return property ? (
      this.renderChildren(value)
    ) : (
      <PagesContext.Consumer>{val => this.renderChildren(val)}</PagesContext.Consumer>
    );
  }
}

ValueComponent.defaultProps = {
  value: '-',
  path: '',
  property: '',
};

ValueComponent.propTypes = {
  value: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  path: PropTypes.string,
  property: PropTypes.string,
};

export const mapStateToProps = (state, props) =>
  props.property ? { value: markdownDatasets.getMetadataValue(state, props) } : {};

export default connect(mapStateToProps)(ValueComponent);
