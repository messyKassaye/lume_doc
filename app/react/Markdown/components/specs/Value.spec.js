import React from 'react';
import { shallow, render } from 'enzyme';

import PagesContext from '../Context';
import { mapStateToProps, ValueComponent } from '../Value.js';
import markdownDatasets from '../../markdownDatasets';

let undefinedValue;

describe('Value', () => {
  it('should render the value passed by mapStateToProps', () => {
    spyOn(markdownDatasets, 'getMetadataValue').and.returnValue('some metadata value');
    const props = { property: 'propValue' };
    const mappedProps = mapStateToProps('state', props);
    const component = shallow(<ValueComponent {...{ ...props, ...mappedProps }} />);

    expect(markdownDatasets.getMetadataValue).toHaveBeenCalledWith('state', {
      property: 'propValue',
    });
    expect(component).toMatchSnapshot();
  });

  it('should render a placeholder when value is "null"', () => {
    spyOn(markdownDatasets, 'getMetadataValue').and.returnValue(undefinedValue);
    const props = { property: 'propValue2' };
    const mappedProps = mapStateToProps('state', props);
    const component = shallow(<ValueComponent {...{ ...props, ...mappedProps }} />);

    expect(markdownDatasets.getMetadataValue).toHaveBeenCalledWith('state', {
      property: 'propValue2',
    });
    expect(component).toMatchSnapshot();
  });

  describe('when using the context', () => {
    it('should render the value in the context path', () => {
      const rendered = render(
        <span>
          <PagesContext.Provider value={{ name: 'Bruce Wayne' }}>
            <ValueComponent path="name" />
          </PagesContext.Provider>
        </span>
      );
      expect(rendered).toMatchSnapshot();
    });
  });
});
