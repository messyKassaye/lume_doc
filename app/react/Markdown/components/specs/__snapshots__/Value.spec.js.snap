// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Value should render a placeholder when value is "null" 1`] = `"-"`;

exports[`Value should render the value passed by mapStateToProps 1`] = `"some metadata value"`;

exports[`Value when using the context should render the value in the context path 1`] = `
<span>
  Bruce Wayne
</span>
`;
